routes/stats.py

```diff
@@ -175,48 +171,196 @@ def stats() -> Union[Response, str]:
+                articles_total["sent"] += r.sent
+                articles_total["parsed"] += r.parsed
+
+            # Gender stats
+            gq = GenderQuery()
+            gender_result = gq.execute(session)
+
+            gender_total = dict(
+                kvk=Decimal(), kk=Decimal(), hk=Decimal(), total=Decimal()
+            )
+            for r in gender_result:
+                gender_total["kvk"] += r.kvk
+                gender_total["kk"] += r.kk
+                gender_total["hk"] += r.hk
+                gender_total["total"] += r.kvk + r.kk + r.hk
+
+            # Author stats
+            author_result = top_authors(session=session)
+
+            # Scraping and parsing stats
+            chart_data = chart_stats(session=session, num_days=days)
+
+            return render_template(
+                "stats.html",
+                title="Tölfræði",
+                articles_result=articles_result,
+                articles_total=articles_total,
+                gender_result=gender_result,
+                gender_total=gender_total,
+                author_result=author_result,
+                scraped_chart_data=json.dumps(chart_data["scraped"]),
+                scraped_avg=int(round(chart_data["scraped"]["avg"])),
+                parsed_chart_data=json.dumps(chart_data["parsed"]),
+                parsed_avg=round(chart_data["parsed"]["avg"], 1),
+            )
+    except Exception as e:
+        logging.error(f"Error rendering stats page: {e}")
+        return Response(f"Error: {e}", status=500)
+
+
+_DEFAULT_QUERY_STATS_PERIOD = 30
+_MAX_QUERY_STATS_PERIOD = 30
+
+
+def query_stats_data(
+    session=None, num_days: int = _DEFAULT_QUERY_STATS_PERIOD
+) -> Dict[str, Any]:
+    """Return all data for query stats dashboard."""
+    today = datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)
+
+    labels = []
+    query_count_data = []
+
+    # Get query count for each day
+    # We change locale to get localized date weekday/month names
+    with changedlocale(category="LC_TIME"):
+        for n in range(0, num_days):
+            days_back = num_days - n - 1
+            start = today - timedelta(days=days_back)
+            end = today - timedelta(days=days_back - 1)
+
+            # Generate date label
+            dfmtstr = "%a %-d. %b"
+            labels.append(start.strftime(dfmtstr))
+
+            # Get query count for day
+            q = list(QueryCountQuery.period(start, end, enclosing_session=session))
+            query_count_data.append(q[0][0])
+
+    query_avg = sum(query_count_data) / num_days
+
+    start = today - timedelta(days=num_days)
+    end = datetime.utcnow()
+
+    # Query types
+    res = QueryTypesQuery.period(
+        start=start,
+        end=end,
+        enclosing_session=session,
+    )
+
+    # This function is used to ensure that all the query
+    # types have a fixed, unique color on the pie chart.
+    def _hexcolor4string(s: str) -> str:
```

Comment:
Quite a hacky solution. A better one would generate an HSL tuple with constant S (saturation) and L (luminance) but different H (hues), then convert it to RGB.

Reply:
Spoken like someone who has addressed a similar problem before! Will do, cap'n! :)

Reply:
Speaking of which, I vaguely remembering solving a similar problem while developing a game for Gogogic in 2010. The point was to create random uniquely coloured shapes that were perceptually different. I used some weird perceptual color space, as I recall. Maybe I still have that code around. That being said, maybe we're overengineering the in-house pie chart.
 ;)